<a href="https://colab.research.google.com/github/davidpalacio1/TP-1-LDD/blob/main/TP_2_LDD_versionDavid.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TP2 - Laboratorio de Datos  
## Verano 2026
## Encuesta permanente de hogares- Argentina  

**Universidad de Buenos Aires**  
**Facultad de Ciencias Exactas y Naturales (FCEyN)**  
<br>


**Integrantes:** Lautaro Herrera - Nicolas Corral - David Palacio Velásquez
<br>
**Fecha:** 9 de marzo de 2026  

<br>


In [1]:
import numpy as np
import pandas as pd


#Limpieza de datos.

Esta tabla contiene muchas variables, y la gran mayor´ıa de ellas son categ´oricas, por ejemplo REGION
(01 = Gran Buenos Aires, 40 = Noroeste, ...). La informaci´on de todas la variables puede verse en
el archivo EPH_registro_3T2025.pdf, en la secci´on “Dise˜no de registros de la base Personas” (que
es la base que estamos usando).
Armamos un dataset de trabajo dejando solo algunas filas y columnas. Los pasos que se proponen a
continuaci´on pueden seguirlos paso a paso o pueden hacer modificaciones si las consideran apropiadas,
por ejemplo utilizar algunas variables m´as o menos de las que proponemos.

##Ejercicio 1

Leer el archivo a un DataFrame (usar sep = ";" dado que el archivo tiene valores separados
por ;)

In [3]:
df = pd.read_csv("usu_individual_T325.txt", sep = ";")

/tmp/ipykernel_374/1016258245.py:1: DtypeWarning: Columns (102) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("usu_individual_T325.txt", sep = ";")


In [4]:
df

,CODUSU,ANO4,TRIMESTRE,NRO_HOGAR,COMPONENTE,H15,REGION,MAS_500,AGLOMERADO,PONDERA,...,V21_03_M,V22_01_M,V22_02_M,V22_03_M,P_DECCF,P_RDECCF,P_GDECCF,P_PDECCF,P_IDECCF,P_ADECCF
0,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,8,0,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
1,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,9,0,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
2,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,1,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
3,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,2,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
4,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,3,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44941,TQRMNOPXUHJOLQCDEHLEH00862602,2025,3,2,5,0,40,N,25,227,...,0,0,0,0,2.0,2.0,NaN,2.0,2.0,2.0
44942,TQRMNOQPQHMKLQCDEHLEH00891262,2025,3,1,1,1,40,N,25,336,...,0,0,0,0,7.0,9.0,NaN,8.0,8.0,10.0
44943,TQRMNOQPQHMKLQCDEHLEH00891262,2025,3,1,2,1,40,N,25,336,...,0,0,0,0,7.0,9.0,NaN,8.0,8.0,10.0
44944,TQRMNOQPQHMKLQCDEHLEH00891262,2025,3,1,3,0,40,N,25,336,...,0,0,0,0,7.0,9.0,NaN,8.0,8.0,10.0


In [5]:
print("Las dimensiones de df son :", df.shape)

Las dimensiones de df son : (44946, 235)


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44946 entries, 0 to 44945
Columns: 235 entries, CODUSU to P_ADECCF
dtypes: float64(148), int64(81), object(6)
memory usage: 80.6+ MB


In [7]:
print("En df tengo las columnas :" , list(df.columns))

En df tengo las columnas : ['CODUSU', 'ANO4', 'TRIMESTRE', 'NRO_HOGAR', 'COMPONENTE', 'H15', 'REGION', 'MAS_500', 'AGLOMERADO', 'PONDERA', 'CH03', 'CH04', 'CH05', 'CH06', 'CH07', 'CH08', 'CH09', 'CH10', 'CH11', 'CH12', 'CH13', 'CH14', 'CH15', 'CH15_COD', 'CH16', 'CH16_COD', 'NIVEL_ED', 'ESTADO', 'CAT_OCUP', 'CAT_INAC', 'IMPUTA', 'PP02C1', 'PP02C2', 'PP02C3', 'PP02C4', 'PP02C5', 'PP02C6', 'PP02C7', 'PP02C8', 'PP02E', 'PP02H', 'PP02I', 'PP03C', 'PP03D', 'PP3E_TOT', 'PP3F_TOT', 'PP03G', 'PP03H', 'PP03I', 'PP03J', 'INTENSI', 'PP04A', 'PP04B_COD', 'PP04B1', 'PP04B2', 'PP04B3_MES', 'PP04B3_ANO', 'PP04B3_DIA', 'PP04C', 'PP04C99', 'PP04D_COD', 'PP04G', 'PP05B2_MES', 'PP05B2_ANO', 'PP05B2_DIA', 'PP05C_1', 'PP05C_2', 'PP05C_3', 'PP05E', 'PP05F', 'PP05H', 'PP06A', 'PP06C', 'PP06D', 'PP06E', 'PP06H', 'PP07A', 'PP07C', 'PP07D', 'PP07E', 'PP07F1', 'PP07F2', 'PP07F3', 'PP07F4', 'PP07F5', 'PP07G1', 'PP07G2', 'PP07G3', 'PP07G4', 'PP07G_59', 'PP07H', 'PP07I', 'PP07J', 'PP07K', 'PP08D1', 'PP08D4', 'PP08F

## Ejercicio 2

Eliminar a todos los individuos que no completaron la encuesta (columna H15).

In [8]:
# Quienes completaron la encuesta tienen un valor 1 en la columna H15 y quienes no la completaron tienen valor 0.
# Para descartar a quienes no completaron la encuesta, me quedo solo con aquellos que tienen valor 1 en ese columna:

df = df[df["H15"] == 1]

In [9]:
df

,CODUSU,ANO4,TRIMESTRE,NRO_HOGAR,COMPONENTE,H15,REGION,MAS_500,AGLOMERADO,PONDERA,...,V21_03_M,V22_01_M,V22_02_M,V22_03_M,P_DECCF,P_RDECCF,P_GDECCF,P_PDECCF,P_IDECCF,P_ADECCF
2,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,1,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
3,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,2,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
4,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,3,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
5,TQRMNOTYPHKOKQCDEHNHB00867020,2025,3,1,4,1,42,S,27,528,...,0,0,0,0,1.0,1.0,1.0,NaN,1.0,1.0
6,TQRMNOQSSHKLKRCDEIHJF00868831,2025,3,1,1,1,44,N,31,140,...,0,0,0,0,3.0,2.0,NaN,4.0,4.0,2.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44939,TQRMNOPXUHJOLQCDEHLEH00862602,2025,3,2,3,1,40,N,25,227,...,0,0,0,0,2.0,2.0,NaN,2.0,2.0,2.0
44940,TQRMNOPXUHJOLQCDEHLEH00862602,2025,3,2,4,1,40,N,25,227,...,0,0,0,0,2.0,2.0,NaN,2.0,2.0,2.0
44942,TQRMNOQPQHMKLQCDEHLEH00891262,2025,3,1,1,1,40,N,25,336,...,0,0,0,0,7.0,9.0,NaN,8.0,8.0,10.0
44943,TQRMNOQPQHMKLQCDEHLEH00891262,2025,3,1,2,1,40,N,25,336,...,0,0,0,0,7.0,9.0,NaN,8.0,8.0,10.0


In [10]:
print("El tamaño inicial de df era (44946, 235), ahora es " , df.shape , ". Es decir, 4947 filas menos.")

El tamaño inicial de df era (44946, 235), ahora es  (39999, 235) . Es decir, 4947 filas menos.


##Ejercicio 3

Seleccionar las siguientes columnas: ”REGION”, ”MAS 500”, ”CH04”, ”CH03”, ”CH06”, ”CH07”,
”CH09”, ”CH10”, ”NIVEL ED”, ”ESTADO”, ”CAT OCUP”, ”CAT INAC”, ”SECTOR”, ”PP02B”,
”PP02C1”, ”PP02C2”, ”PP02C3”, ”PP02C4”, ”PP02C5”, ”PP02C6”, ”PP02C7”, ”PP02C8”,
”PP02D”, ”PP02D”, ”PP02F”, ”PP02G”, ”PP02H”, ”PP02I”, ”PP03C”, ”PP03D”, ”PP3E TOT”,
”PP3F TOT”, ”PP03G”, ”PP03H”, ”PP04A”, ”PP04A1”, ”PP04A1”, ”PP04B1”, ”PP04B2”,
”PP04B3 ANO”, ”PP04C”, ”PP03I”, ”PP03J”, ”PP03K”, ”INTENSI”, ”PP04G”, ”P47T”

In [11]:
df.columns[df.columns.duplicated()]

Index([], dtype='object')

In [12]:
columnas_CORREGIBLE = [
    "REGION", "MAS_500", "CH04", "CH03", "CH06", "CH07",
"CH09", "CH10", "NIVEL_ED", "ESTADO", "CAT_OCUP", "CAT_INAC", "SECTOR", "PP02B",
"PP02C1", "PP02C2", "PP02C3", "PP02C4", "PP02C5", "PP02C6", "PP02C7", "PP02C8",
"PP02D", "PP02D", "PP02F", "PP02G", "PP02H", "PP02I", "PP03C", "PP03D", "PP3E_TOT",
"PP3F_TOT", "PP03G", "PP03H", "PP04A", "PP04A1", "PP04A1", "PP04B1", "PP04B2",
"PP04B3_ANO", "PP04C", "PP03I", "PP03J", "PP03K", "INTENSI", "PP04G", "P47T"
]


In [13]:
# Más adelante, en el Ejercicio 4, obtengo un error al usar esta lista. Al revisar, vi que hay valores repetidos.
# Para verificar:

print("La cantidad de elementos en columnas_CORREGIBLE es:" , len(columnas_CORREGIBLE))
print("La cantidad de elementos en columnas_CORREGIBLE sin contar repetidos es:" , len(set(columnas_CORREGIBLE))) # set() elimina los elementos repetidos. Así, len(columnas_CORREGIBLE) dice el tamaño sin repetidos.


print("¿Hay columnas repetidas? :" , len(columnas_CORREGIBLE) != len(set(columnas_CORREGIBLE)))

La cantidad de elementos en columnas_CORREGIBLE es: 47
La cantidad de elementos en columnas_CORREGIBLE sin contar repetidos es: 45
¿Hay columnas repetidas? : True


In [14]:
# Entonces, saco los valores repetidos. Para esto no me sirve set() pues, aunque saca los repetidos, no mantiene el orden:

columnas = list(dict.fromkeys(columnas_CORREGIBLE))

In [15]:
columnas

['REGION',
 'MAS_500',
 'CH04',
 'CH03',
 'CH06',
 'CH07',
 'CH09',
 'CH10',
 'NIVEL_ED',
 'ESTADO',
 'CAT_OCUP',
 'CAT_INAC',
 'SECTOR',
 'PP02B',
 'PP02C1',
 'PP02C2',
 'PP02C3',
 'PP02C4',
 'PP02C5',
 'PP02C6',
 'PP02C7',
 'PP02C8',
 'PP02D',
 'PP02F',
 'PP02G',
 'PP02H',
 'PP02I',
 'PP03C',
 'PP03D',
 'PP3E_TOT',
 'PP3F_TOT',
 'PP03G',
 'PP03H',
 'PP04A',
 'PP04A1',
 'PP04B1',
 'PP04B2',
 'PP04B3_ANO',
 'PP04C',
 'PP03I',
 'PP03J',
 'PP03K',
 'INTENSI',
 'PP04G',
 'P47T']

In [16]:
# Si tengo un montón de palabras de la forma ”REGION”, con ” ” en vez de " ", ¿Que código puedo usar para que todos los ” se reemplazen automáticamente por "?

In [17]:
df = df[columnas]
df

,REGION,MAS_500,CH04,CH03,CH06,CH07,CH09,CH10,NIVEL_ED,ESTADO,...,PP04B1,PP04B2,PP04B3_ANO,PP04C,PP03I,PP03J,PP03K,INTENSI,PP04G,P47T
2,42,S,2,1,38,1,1,2,1,3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,552000.0
3,42,S,1,2,39,1,1,2,4,1,...,2.0,0.0,0.0,1.0,2.0,2.0,0.0,2.0,8.0,400000.0
4,42,S,1,3,18,5,1,1,1,3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
5,42,S,1,3,17,5,1,2,3,3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
6,44,N,1,1,50,2,1,2,3,1,...,2.0,0.0,0.0,10.0,2.0,2.0,0.0,2.0,11.0,1200000.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44939,40,N,1,3,13,5,1,1,3,3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
44940,40,N,2,3,12,5,1,1,1,3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
44942,40,N,1,1,46,1,1,2,6,1,...,2.0,0.0,0.0,8.0,2.0,2.0,0.0,2.0,11.0,1370000.0
44943,40,N,2,2,38,1,1,2,6,1,...,2.0,0.0,0.0,3.0,1.0,1.0,2.0,1.0,11.0,700000.0


##Ejercicio 4

Muchas de estas columnas contienen datos faltantes, en general por preguntas que no co-
rresponden. Por ejemplo, preguntas sobre ocupaci´on a personas desocupadas. Para las si-
guientes columnas, reemplazar los datos faltantes por 0: ”EMPLEO”, ”PP03C”, ”PP03D”,
”PP3E TOT”, ”PP3F TOT”, ”PP03G”, ”PP03H”, ”PP03I”, ”PP03J”, ”PP03K”, ”PP04B1”,
”PP04B2”, ”PP04B3 ANO”, ”PP04C”, ”PP04A”, ”PP04A1”, ”PP04G”, ”INTENSI” (la ma-
yor´ıa son variables categ´oricas que van a ser reemplazadas por dummies para las distintas
categor´ıas).

In [18]:
columnas_a_reemplazar = [
    "EMPLEO",
    "PP03C", "PP03D","PP3E_TOT","PP03G", "PP03H", "PP03I", "PP03J", "PP03K", "PP04B1","PP04B2", "PP04B3_ANO", "PP04C", "PP04A", "PP04A1", "PP04G", "INTENSI" ]

In [21]:
df[columnas_a_reemplazar] = df[columnas_a_reemplazar].fillna(0)

KeyError: "['EMPLEO'] not in index"

In [20]:
# No corre porque "EMPLEO" no está entre las columas seleccionadas en el Ejercicio 3. Dos opciones:
        # 1. Agregar "EMPLEO" a las columnas seleccionadas en 3.
        # 2. No tener en cuenta a "EMPLEO" para las columnas seleccionadas y sacarlo de las columnas a reemplazar.

In [22]:
# Opción 1. Agregando "EMPLEO" a las columnas seleccionadas:

df = pd.read_csv("usu_individual_T325.txt", sep = ";")
df = df[df["H15"] == 1] # Hasta acá, lo hecho hasta 2.

# Agrego a "EMPLEO":

columnas = [
    "EMPLEO",
  'REGION', 'MAS_500', 'CH04', 'CH03', 'CH06', 'CH07', 'CH09', 'CH10', 'NIVEL_ED', 'ESTADO', 'CAT_OCUP', 'CAT_INAC', 'SECTOR', 'PP02B', 'PP02C1', 'PP02C2', 'PP02C3', 'PP02C4', 'PP02C5', 'PP02C6', 'PP02C7', 'PP02C8', 'PP02D', 'PP02F', 'PP02G', 'PP02H', 'PP02I', 'PP03C', 'PP03D', 'PP3E_TOT', 'PP3F_TOT', 'PP03G', 'PP03H', 'PP04A', 'PP04A1', 'PP04B1', 'PP04B2', 'PP04B3_ANO', 'PP04C', 'PP03I','PP03J','PP03K','INTENSI','PP04G','P47T']

df = df[columnas]

/tmp/ipykernel_374/2412743933.py:3: DtypeWarning: Columns (102) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("usu_individual_T325.txt", sep = ";")


In [23]:
# Las columnas donde voy a reemplazar:

columnas_a_reemplazar = [
    "EMPLEO",
    "PP03C", "PP03D","PP3E_TOT", "PP3F_TOT", "PP03G", "PP03H", "PP03I", "PP03J", "PP03K", "PP04B1","PP04B2", "PP04B3_ANO", "PP04C", "PP04A", "PP04A1", "PP04G", "INTENSI" ]

In [24]:
# Reemplazo los NA con 0:

df[columnas_a_reemplazar] = df[columnas_a_reemplazar].fillna(0)

In [25]:
# Opción 2. No teniendo en cuenta a "EMPLEO" para las columnas seleccionadas y sacandolo de las columnas a reemplazar:

# df = pd.read_csv("usu_individual_T325.txt", sep = ";")
# df = df[df["H15"] == 1] # Hasta acá, lo hecho hasta 2.

# columnas sin "EMPLEO":

# columnas = ['REGION', 'MAS_500', 'CH04', 'CH03', 'CH06', 'CH07', 'CH09', 'CH10', 'NIVEL_ED', 'ESTADO', 'CAT_OCUP', 'CAT_INAC', 'SECTOR', 'PP02B', 'PP02C1', 'PP02C2', 'PP02C3', 'PP02C4', 'PP02C5', 'PP02C6', 'PP02C7', 'PP02C8', 'PP02D', 'PP02F', 'PP02G', 'PP02H', 'PP02I', 'PP03C', 'PP03D', 'PP3E_TOT', 'PP3F_TOT', 'PP03G', 'PP03H', 'PP04A', 'PP04A1', 'PP04B1', 'PP04B2', 'PP04B3_ANO', 'PP04C', 'PP03I','PP03J','PP03K','INTENSI','PP04G','P47T']

# df = df[columnas]

# Las columnas donde voy a reemplazar, sin "EMPLEO":

# columnas_a_reemplazar = ["PP03C", "PP03D","PP3E_TOT","PP03G", "PP03H", "PP03I", "PP03J", "PP03K", "PP04B1","PP04B2", "PP04B3_ANO", "PP04C", "PP04A", "PP04A1", "PP04G", "INTENSI" ]

# Reemplazo los NA con 0:

# df[columnas_a_reemplazar] = df[columnas_a_reemplazar].fillna(0)

In [26]:
# Avanzo usando la versión de Opción 1. Intercambiable por la 2, la dejo marcada con #.

##Ejercicio 5

Eliminar las filas que no tienen valor en la columna de ingreso total (”P47T”).

In [27]:
# Revisamos la cantidad de valores  fantantes (es decir, los que están como NA) en P47T:

df["P47T"].isna().sum()

np.int64(0)

No encuentramos valores faltantes en  la columna de ingreso total P47T. Por lo tanto no eliminamos observaciones en este paso.

##Ejercicio 6

Eliminar filas con datos faltantes.

In [28]:

# Verificamos si hay filas con datos faltantes en df:

print("En df se encuentran", df.isna().sum().sum(), "filas con datos faltantes.")

En df se encuentran 19933 filas con datos faltantes.


In [29]:
# Identificamos las columnas con NA:

df.isna().sum()[df.isna().sum() > 0]

,0
SECTOR,19933


La única variable con valores faltantes es SECTOR. Hay 19.933 filas con NaN en SECTOR. El resto de las variables no presenta valores faltantes.

En df tenemos 39.999 filas. Así, las 19.933 filas con valores faltantes en SECTOR corresponden a cerca del 50% de la muestra. Antes de eliminar estas filas, estudiamos un poco más la situación.

Según el documento oficial “Diseño de registros y estructura para las bases preliminares Hogar y Personas” (INDEC - 3° trimestre de 2025), en la sección correspondiente a la base Personas, se tiene que la variable SECTOR clasifica el sector de inserción laboral de los ocupados. Puede tomar los siguientes valores:

        - Formal = 1
        - Informal = 2
        - Hogares = 3
        - Ns/Nr = 9

En la misma sección se define la variable ESTADO, que toma los valores:

        - Ocupado = 1
        - Desocupado = 2
        - Inactivo = 3

De esta definición se desprende que SECTOR solo es conceptualmente aplicable a las personas que estén ocupadas (ESTADO = 1). No tiene sentido económico ni estadístico asignar un sector laboral a desocupados o inactivos. Por lo tanto, la ausencia de valor en SECTOR para estos casos no constituye un error de registro, sino una consecuencia lógica del diseño del cuestionario.

Así, los NA presentes no parecen ser erores de carga, omisiones accidentales o inconsistencias del dataset. En consecuencia, estos NaN en personas desocupadas o inactivas no constituye un error en calidad de los datos.

Eliminar estas observaciones implicaría descartar aproximadamente la mitad del dataset, introduciendo un sesgo en la composición de la muestra.

No consideramos apropiado eliminar estas filas en este paso.

##Ejercicio 7

Convertir las siguientes variables categóricas a dummies: ”ESTADO”, ”REGION”, ”MAS_500”,
”CAT_OCUP”, ”CAT_INAC”, ”SECTOR”, ”PP02B”, ”PP02C1”, ”PP02C2”, ”PP02C3”, ”PP02C4”,
”PP02C5”, ”PP02C6”, ”PP02C7”, ”PP02C8”, ”PP02D”, ”PP02D”, ”PP02F”, ”PP02G”, ”PP02H”,
”PP02I”, ”PP03C”, ”PP03D”, ”PP3E_TOT”, ”PP3F_TOT”, ”PP03G”, ”PP03H”, ”PP03I”,
”PP03J”, ”PP03K”, ”INTENSI”, ”PP04A”, ”PP04A1”, ”PP04A1”, ”PP04G”. Pueden usar
Formulaic o pd.get_dummies.


Nota: La variable SECTOR también es categórica pero vamos a utilizarla para clasificación
por eso no la convertimos a dummies

In [30]:
lista = [
    "ESTADO", "REGION", "MAS_500", "CAT_OCUP", "CAT_INAC", "SECTOR",
    "PP02B", "PP02C1", "PP02C2", "PP02C3", "PP02C4", "PP02C5",
    "PP02C6", "PP02C7", "PP02C8", "PP02D", "PP02D", "PP02F",
    "PP02G", "PP02H", "PP02I", "PP03C", "PP03D", "PP3E_TOT",
    "PP3F_TOT", "PP03G", "PP03H", "PP03I", "PP03J", "PP03K",
    "INTENSI", "PP04A", "PP04A1", "PP04A1", "PP04G"
]

In [31]:
# Las variables dadas en el enunciado, sacando los duplicados PP02D y PP04A1:

variables_enunciado = [
    "ESTADO", "REGION", "MAS_500",
    "CAT_OCUP", "CAT_INAC", "SECTOR",
    "PP02B", "PP02C1", "PP02C2", "PP02C3", "PP02C4",
    "PP02C5", "PP02C6", "PP02C7", "PP02C8",
    "PP02D", "PP02F", "PP02G", "PP02H", "PP02I",
    "PP03C", "PP03D", "PP3E_TOT", "PP3F_TOT",
    "PP03G", "PP03H", "PP03I", "PP03J", "PP03K",
    "INTENSI", "PP04A", "PP04A1", "PP04G"
]

# Revisamos los tipos de las variables:

df[variables_enunciado].dtypes

,0
ESTADO,int64
REGION,int64
MAS_500,object
CAT_OCUP,int64
CAT_INAC,int64
SECTOR,float64
PP02B,int64
PP02C1,int64
PP02C2,int64
PP02C3,int64


In [32]:
df[variables_enunciado].dtypes.value_counts()

,count
int64,18
float64,14
object,1


In [33]:


tipos = df[variables_enunciado].dtypes

{
    tipo: list(tipos[tipos == tipo].index)
    for tipo in tipos.unique()
}

{dtype('int64'): ['ESTADO',
  'REGION',
  'CAT_OCUP',
  'CAT_INAC',
  'PP02B',
  'PP02C1',
  'PP02C2',
  'PP02C3',
  'PP02C4',
  'PP02C5',
  'PP02C6',
  'PP02C7',
  'PP02C8',
  'PP02D',
  'PP02F',
  'PP02G',
  'PP02H',
  'PP02I'],
 dtype('O'): ['MAS_500'],
 dtype('float64'): ['SECTOR',
  'PP03C',
  'PP03D',
  'PP3E_TOT',
  'PP3F_TOT',
  'PP03G',
  'PP03H',
  'PP03I',
  'PP03J',
  'PP03K',
  'INTENSI',
  'PP04A',
  'PP04A1',
  'PP04G']}

Tenemos:

-  De tipo Int ('int64'): 'ESTADO','REGION','CAT_OCUP', 'CAT_INAC', 'PP02B', 'PP02C1', 'PP02C2', 'PP02C3', 'PP02C4', 'PP02C5', 'PP02C6', 'PP02C7', 'PP02C8', 'PP02D', 'PP02F', 'PP02G', 'PP02H', 'PP02I'.

- De tipo Object ('O'): 'MAS_500'.

- De tipo Float ('float64'): 'SECTOR', 'PP03C', 'PP03D', 'PP3E_TOT', 'PP3F_TOT', 'PP03G', 'PP03H', 'PP03I', 'PP03J', 'PP03K', 'INTENSI', 'PP04A', 'PP04A1', 'PP04G'.

Estas pueden clasificarse como:

  1. **Categóricas nominales**: Representan categorías sin un orden natural
   Por ejemplo ESTADO, REGION, CAT_OCUP, PP02B y PP04A. En ellas los valores numéricos representan códigos de categoría, no magnitudes.


  2. **Categóricas ordinales**: Poseen cierto orden conceptual entre las categorías, aunque la distancia entre ellas no necesariamente es uniforme ni interpretable como una magnitud cuantitativa. Por ejemplo PP03D (cantidad de ocupaciones) o INTENSI (subocupado, pleno, sobreocupado). Si bien hay un orden, tratar estas variables como numéricas implicaría asumir que la diferencia entre categorías es constante, lo cual no está garantizado. Por ello, convertirlas en dummies resulta adecuado.
  

  3. **Variables cuantitativas**: Representan magnitudes medibles en una escala numérica con significado. Tenemos a PP3E_TOT (total de horas trabajadas en la semana en la ocupación principal)  y PP3F_TOT(total de horas trabajadas en otras ocupaciones). La distancia entre sus valores es interpretable, nos permite comparar diferencias, por ejemplo trabajar 40 horas vs trabajar 20 horas.

Aunque el enunciado incluye a PP3E_TOT y PP3F_TOT dentro del listado de variables categóricas, transformarlas en dummies implicaría perder información relevante. Por lo tanto, decidimos  mantenerlas como variables numéricas y no transformarlas.



Como en el Ejercicio 3, la lista de variables contiene duplicados. Eliminamos esos duplicados y excluimos a SECTOR, que se usará más adelante, y a PP3E_TOT y PP3F_TOT.

In [34]:
# Las variables categóricas a transformar:

variables_categoricas = [
    "ESTADO", "REGION", "MAS_500", "CAT_OCUP", "CAT_INAC", "PP02B", "PP02C1", "PP02C2", "PP02C3", "PP02C4", "PP02C5", "PP02C6", "PP02C7", "PP02C8",
    "PP02D", "PP02F", "PP02G", "PP02H", "PP02I", "PP03C", "PP03D", "PP03G", "PP03H", "PP03I", "PP03J", "PP03K", "INTENSI", "PP04A", "PP04A1", "PP04G"
]

In [35]:
# Convertimos las variables categóricas a dummies. Usamos pd.get_dummies:
# Se usa drop_first=True para evitar multicolinealidad perfecta. (OBSERVACIÓN: Ni idea por qué es esto, busqué como usar get_dummies y encontré eso. Tengo que revisar bien)

df = pd.get_dummies(
    df,
    columns=variables_categoricas,
    drop_first=True
)

In [36]:
df.head()

,EMPLEO,CH04,CH03,CH06,CH07,CH09,CH10,NIVEL_ED,SECTOR,PP3E_TOT,...,PP04G_4.0,PP04G_5.0,PP04G_6.0,PP04G_7.0,PP04G_8.0,PP04G_9.0,PP04G_10.0,PP04G_11.0,PP04G_12.0,PP04G_13.0
2,0.0,2,1,38,1,1,2,1,NaN,0.0,...,False,False,False,False,False,False,False,False,False,False
3,2.0,1,2,39,1,1,2,4,2.0,36.0,...,False,False,False,False,True,False,False,False,False,False
4,0.0,1,3,18,5,1,1,1,NaN,0.0,...,False,False,False,False,False,False,False,False,False,False
5,0.0,1,3,17,5,1,2,3,NaN,0.0,...,False,False,False,False,False,False,False,False,False,False
6,1.0,1,1,50,2,1,2,3,1.0,35.0,...,False,False,False,False,False,False,False,True,False,False


##Ejercicio 8

Para las siguientes variables realizar las modificaciones propuestas:

(a) Reemplazar CH07 (estado civil) por una variable que indique si es soltero o no (para no usar tantas dummies).

(b) Reemplazar CH09 por una variable que indique si sabe leer.

(c) Eliminar a los individuos que no respondieron CH09 (cantidad de personas que trabajan
en el lugar de trabajo), corresponden a código 99.

(d) En nivel educativo, reemplazar 7 por 0.

(e) Llamar df_clean al DataFrame resultante.

Comenzamos generando una copia del df original, que llamaremos df_clean (como se indica en el punto (c)), y aplicamos los cambios en ese dataset.

In [37]:
df_clean = df.copy()

(a) La variable CH07 indica el estado civil, donde sus valores representan:

 - 1 = Unido

  - 2 = Casado

  - 3 = Separado/Divorciado

  - 4 = Viudo

  - 5 = Soltero

Lo reemplazamos una variable binaria "SOLTERO", con 1 si es soltero y 0 si no lo es.

In [38]:
# Reemplazamos por 1 si es soltero (codigo 5) y 0 en otro caso:

df_clean["SOLTERO"] = (df_clean["CH07"] == 5).astype(int)

# Eliminamos a CH07:

df_clean = df_clean.drop(columns=["CH07"])

(b) La variable CH09 indica si la persona sabe leer y tiene tres valores:

  - 1	= Sí
  - 2	= No
  - 3	= Menor de 2 años

La reemplazamos por una variable binaria "LEE", con 1 si sabe leer y 0 si no. Los menores de 2 años no saben leer, por lo que los agrupamos con el valor 0.

In [39]:
# Variable que indica si sabe leer
df_clean["LEE"] = (df_clean["CH09"] == 1).astype(int)

# Eliminamos la variable original
df_clean = df_clean.drop(columns=["CH09"])

(c) El enunciado pide eliminar los casos con valor 99 en la variable CH09, interpretados como individuos que no respondieron la pregunta sobre la cantidad de personas que trabajan en el lugar de trabajo.

Pero, según la documentación de la Encuesta Permanente de Hogares, CH09 corresponde a la variable "sabe leer y esribir", cuyos valores posibles son 1 (sí), 2 (no) y 3 (menor de 2 años), la cual se transformó en el punto anterior y no tiene valor 99.

La variable que registra la cantidad de personas que trabajan en el lugar de trabajo es PP04C, donde el valor 99 indica "No sabe / No responde".

En consecuencia, decidimos aplicar el enunciado a esta variable y eliminar las observaciones con PP04C = 99, que se corresponden con los individuos que no respondieron en ese caso.

In [41]:
# Revisamos cuantos individuos tienen valor 99 en PP04C:

(df_clean["PP04C"] == 99).sum()

np.int64(2687)

In [43]:
# Eliminamos:

df_clean = df_clean[df_clean["PP04C"] != 99]

Se encontraron 2687 observaciones con valor 99, las cuales fueron eliminadas.

(d) En la EPH, el nivel educativo está dado por NIVEL_ED, cuyos valores son:

  - 1 = Primario incompleto (incluye educación especial)
  - 2 = Primario completo
  - 3 = Secundario incompleto
  - 4 = Secundario completo
  - 5 = Superior y universitario incompleto
  - 6 = Superior y universitario completo
  - 7 = Sin instrucción
  - 9 = Ns/Nr

Reemplazamos los valores con 7 por 0.

In [45]:
# Primero, revisamos cuantos datos hay para cada valor:

df_clean["NIVEL_ED"].value_counts()

,count
NIVEL_ED,
4,9049
3,8792
6,5618
5,5241
2,4816
1,3559
7,237


In [47]:
# Hay 237 datos con valor 7. Los eliminamos:

df_clean["NIVEL_ED"] = df_clean["NIVEL_ED"].replace(7, 0)

df_clean["NIVEL_ED"].value_counts()

,count
NIVEL_ED,
4,9049
3,8792
6,5618
5,5241
2,4816
1,3559
0,237


# Clustering

Nuestro primer objetivo es agrupar a los individuos en grupos diferenciados. En este caso, no tenemos
una respuesta correcta que queremos predecir. Queremos realizar un agrupamiento y ver si podemos
identificar características de cada grupo.


## Ejercicio 9

Generar un DataFrame df_clustering pasando la variable SECTOR a dummies.


En el dataset limpio df_clean, la variable SECTOR indica el tipo de trabajo del individuo, donde sus valores son:

  - 1 = Trabajo formal

  - 2 = Trabajo informal

  - 3 = Servicio doméstico

  - 9 = No sabe, no responde

Para aplicar clustering necesitamos que los datos sean completamente numéricos. Dado que SECTOR es una variable categórica, la transformamos a dummies.

In [49]:
# Revisamos los valores en SECTOR:

df_clean["SECTOR"].value_counts()

,count
SECTOR,
1.0,10967
2.0,4540
3.0,1413
9.0,459


In [50]:
# En el paso anterior verificamos que no hay valores faltantes. Pasamos a dummies:

df_clustering = pd.get_dummies(df_clean, columns=["SECTOR"])

In [51]:
# Verificamos el DataFrame. El paso anterior debe generar una nueva columna por cada categoría de SECTOR:

df_clustering.head()

,EMPLEO,CH04,CH03,CH06,CH10,NIVEL_ED,PP3E_TOT,PP3F_TOT,PP04B1,PP04B2,...,PP04G_10.0,PP04G_11.0,PP04G_12.0,PP04G_13.0,SOLTERO,LEE,SECTOR_1.0,SECTOR_2.0,SECTOR_3.0,SECTOR_9.0
2,0.0,2,1,38,2,1,0.0,0.0,0.0,0.0,...,False,False,False,False,0,1,False,False,False,False
3,2.0,1,2,39,2,4,36.0,0.0,2.0,0.0,...,False,False,False,False,0,1,False,True,False,False
4,0.0,1,3,18,1,1,0.0,0.0,0.0,0.0,...,False,False,False,False,1,1,False,False,False,False
5,0.0,1,3,17,2,3,0.0,0.0,0.0,0.0,...,False,False,False,False,1,1,False,False,False,False
6,1.0,1,1,50,2,3,35.0,0.0,2.0,0.0,...,False,True,False,False,0,1,True,False,False,False


Obtuvimos las nuevas columnas: SECTOR_1.0	SECTOR_2.0	SECTOR_3.0	SECTOR_9.0.

In [53]:
# Verificamos que solo tengamos variables numéricas (o de tipo bool, equivalentes a 1 y 0):

df_clustering.dtypes.value_counts()

,count
bool,101
float64,8
int64,7


In [54]:
# Antes de continuar con el clustering, revisamos si hay datos faltantes:

df_clustering.isna().sum().sum()

np.int64(0)

In [56]:
# No se encontraron datos faltantes. Las dimensiones del dataset que tenemos son:

print("Dimensiones del conjunto de datos:" , df_clustering.shape)

Dimensiones del conjunto de datos: (37312, 116)


Las dimensiones del conjunto de datos son (37312, 116), es decir que tenemos 116 variables.

## Ejercicio 10

Visualización: Calcular las dos primeras componentes principales Z1 y Z2 del DataFrame df_clustering y realizar un gráfico de dispersión de Z1 vs. Z2. ¿Pueden observar agrupamientos en esta visualización?

Como se encontró en el ejercicio anterior, contamos con una cantidad grande de variables. Esto puede generar problemas de dimensionalidad. Usaremos Análisis de Componentes Principales (PCA), una técnica de reducción de dimensionalidad que permite proyectar los datos en un espacio de menor dimensión conservando la mayor cantidad posible de variabilidad.

Para esto, vamos a calcular las dos primeras componentes principales:

  - Z1: Primera componente principal

  - Z2: Segunda componente principal

En el dataset tenemos variables con escalas muy diferentes. Por ejemplo:

- La edad puede tomar valores entre 0 y 100

- Las variables dummy toman valores de 0 o 1

Antes de aplicar PCA vamos a normalizar las variables, para evitar qué aquellas con mayor escala dominen en el cálculo de las componentes principales. Utilizamos StandardScaler, que transforma cada variable para que tenga media 0 y desviación estándar 1.

In [57]:
# Normalizamos:

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaled = scaler.fit_transform(df_clustering)

In [58]:
pd.DataFrame(X_scaled).describe().loc[["mean", "std"]]

,0,1,2,3,4,5,6,7,8,9,...,106,107,108,109,110,111,112,113,114,115
mean,5.560637e-17,-1.904328e-19,2.989794e-17,-1.030241e-16,4.513257e-17,1.112127e-16,-6.017675e-17,1.447289e-17,1.729130e-16,1.942414e-17,...,2.818405e-17,-9.369292e-17,1.447289e-17,6.855580e-18,-1.572975e-16,-2.856491e-16,-3.046924e-17,2.285193e-17,-5.341639e-17,-1.990022e-17
std,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,...,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00,1.000013e+00


Tras normalizar, los resultados muestran que las medias de las variables son cercanas a 0 y las desviaciones estándar cercanas a 1. Las pequeñas diferencias respecto a estos valores se deben a errores de precisión numérica en los cálculos. Luego, la normalización se realizó correctamente.

Ahora aplicamos Análisis de Componentes Principales:

In [60]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)

pca.fit(X_scaled)

Z = pca.transform(X_scaled)

# Clasificaciòn



# Regresión

Queremos ahora predecir el sueldo de las personas ocupadas.

##Ejercicio 18

Generar un DataFrame df_regresion que contenga solo las personas ocupadas, con ingreso
total positivo (variable P47T).

##Ejercicio 19

Reemplazar la variable P47T por su logaritmo

##Ejercicio 20

Proponer tres modelos distintos para ajustar la variable P47T transformada (pueden incluir
distintas variables, regresi´on ordinaria, Ridge, redes neuronales, ...).
3

##Ejercicio 21

Al menos uno de los modelos propuestos debe depender de hiperpar´ametros. Seleccionar los
hiperpar´ametros por alguna metodolog´ıa apropiada.

##Ejercicio 22

Seleccionar el mejor de los 3 modelos mediante un esquema de validaci´on apropiado, e indicar
el RSME en un conjunto de datos de testeo no utilizado previamente.

##Ejercicio 23

Seleccionar tres variables que consideren interesantes para evaluar la influencia de las mismas
en el salario de una persona. Realizar alg´un experimento simple para ver la influencia seg´un el
modelo seleccionado en el ´ıtem anterior.

Nota: En modelos lineales podr´ıa simplemente observarse el coeficiente correspondiente. En
otros modelos no se puede hacer este an´alisis pero podemos analizar por ejemplo que sucede
cuando modificamos una variable. Por ejemplo, si queremos analizar el efecto del Sexo, podemos
comparar el salario promedio cuando reemplazamos en todas nuestras observaciones el sexo por
varon y el salario promedio reemplazando el sexo por mujer.